src/wsproxy.ml

```diff
@@ -15,124 +15,115 @@
-    (fun () ->
-       let thread1 =
-         lwt_fd_enumerator localfd (realframe (writer (really_write fd) "thread1")) >>= fun _ ->
-         Lwt.return () in
-       let thread2 =
-         lwt_fd_enumerator fd (realunframe (writer (really_write localfd) "thread2")) >>= fun _ ->
-         Lwt.return () in
-       ignore_exn (fun () -> Lwt.join [thread1; thread2]) ()
-    )
-    (fun _ ->
-       ignore_exn (fun () -> Lwt_unix.close fd) () >>= fun () ->
-       ignore_exn (fun () -> Lwt_unix.close localfd) ())
-  >>= fun () ->
-  Printf.printf "FD %s pipe closed: %b %b\n" fdno Lwt_unix.(state fd == Closed) Lwt_unix.(state localfd == Closed)
-  |> Lwt.return
+      Lwt_log.debug_f "unknown protocol" >>= fun () ->
+      Lwt.return (wsframe, wsunframe) 
+  end >>= fun (frame,unframe) ->
+  with_open_connection_fd "localhost" localport ~callback:(fun localfd ->
+      let session_id = Uuidm.v `V4 |> Uuidm.to_string in
+      Lwt_log.debug_f "Starting proxy session %s" session_id >>= fun () ->
+      let thread1 =
+        lwt_fd_enumerator localfd (frame (writer (really_write fd) "thread1")) >>= fun _ ->
+        Lwt.return_unit in
+      let thread2 =
+        lwt_fd_enumerator fd (unframe (writer (really_write localfd) "thread2")) >>= fun _ ->
+        Lwt.return_unit in
+      (* closing the connection in one of the threads above in general leaves the other pending forever,
+       * by using choose here, we make sure that as soon as one of the threads completes, both are closed *)
+      Lwt.choose [thread1; thread2]
+      >>= fun () -> Lwt_log.debug_f "Closing proxy session %s" session_id)
+
 
 let handler sock msg =
-  Lwt_io.printf "Got msg: %s\n" msg >>= fun _ ->
+  Lwt_log.debug_f "Got msg: %s" msg >>= fun () ->
   match Re_str.split (Re_str.regexp "[:]") msg with
   | [protocol;ty;sport] ->
     let port = int_of_string sport in
     proxy sock protocol ty port
   | [protocol;sport] ->
     let port = int_of_string sport in
     proxy sock protocol "hybi10" port
-  | _ ->
-    let sid = Printf.sprintf "%d" (Obj.magic (Lwt_unix.unix_file_descr sock)) in
-    ignore_exn (fun () -> Lwt_unix.close sock) () >>= fun () ->
-    Printf.printf "Malformed message. Socket %s closed: %b\n" sid Lwt_unix.(state sock == Closed)
-    |> Lwt.return
+  | _ -> Lwt.return_unit
+
 
-let _ =
-  if Array.length Sys.argv > 1
-  then Lwt.return (Websockets.runtest ())
-  else
-    begin
-      Lwt_daemon.daemonize ~stdout:`Dev_null ~stdin:`Close ~stderr:`Dev_null ();
-      let filename = "/var/run/wsproxy.pid" in
-      (try Unix.unlink filename with _ -> ());
-      Lwt_main.run begin
-        let pid = Unix.getpid () in
-        Lwt_io.with_file filename ~mode:Lwt_io.output (fun chan ->
-            Lwt_io.fprintf chan "%d" pid) >>= fun _ ->
-        start "wsproxy" handler
-      end
-    end
+let _ = 
+  (* Enable logging for all levels *)
+  Lwt_log.Section.set_level Lwt_log.Section.main Lwt_log.Debug;
+  Lwt_daemon.daemonize ~stdin:`Close ();
```

Comment:
Should we still close stdout & stderr as we're not using these now, we send the debug messages to the syslog using `Lwt_log`?

Reply:
I have left them open in case something is writing on stdout or stderr and we don't know (or don't expect it). In this way it will appear on the logs

Reply:
It's fairly common to reopen stdin/out/err as /dev/null, as that prevents the integer fds being reused. That in turn prevents screw-ups like printing log lines to, or trying to read from, some random connection.